src/languages/swift.js

```diff
@@ -1,65 +1,66 @@
 export default function(hljs) {
-  const SWIFT_KEYWORDS = {
-      // override the pattern since the default of of /\w+/ is not sufficient to
-      // capture the keywords that start with the character "#"
-      $pattern: /[\w#]+/,
-      keyword: '#available #colorLiteral #column #else #elseif #endif #file ' +
-        '#fileLiteral #function #if #imageLiteral #line #selector #sourceLocation ' +
-        '_ __COLUMN__ __FILE__ __FUNCTION__ __LINE__ Any as as! as? associatedtype ' +
-        'associativity break case catch class continue convenience default defer deinit didSet do ' +
-        'dynamic dynamicType else enum extension fallthrough false fileprivate final for func ' +
-        'get guard if import in indirect infix init inout internal is lazy left let ' +
-        'mutating nil none nonmutating open operator optional override postfix precedence ' +
-        'prefix private protocol Protocol public repeat required rethrows return ' +
-        'right self Self set some static struct subscript super switch throw throws true ' +
-        'try try! try? Type typealias unowned var weak where while willSet',
-      literal: 'true false nil',
-      built_in: 'abs advance alignof alignofValue anyGenerator assert assertionFailure ' +
-        'bridgeFromObjectiveC bridgeFromObjectiveCUnconditional bridgeToObjectiveC ' +
-        'bridgeToObjectiveCUnconditional c compactMap contains count countElements countLeadingZeros ' +
-        'debugPrint debugPrintln distance dropFirst dropLast dump encodeBitsAsWords ' +
-        'enumerate equal fatalError filter find getBridgedObjectiveCType getVaList ' +
-        'indices insertionSort isBridgedToObjectiveC isBridgedVerbatimToObjectiveC ' +
-        'isUniquelyReferenced isUniquelyReferencedNonObjC join lazy lexicographicalCompare ' +
-        'map max maxElement min minElement numericCast overlaps partition posix ' +
-        'precondition preconditionFailure print println quickSort readLine reduce reflect ' +
-        'reinterpretCast reverse roundUpToAlignment sizeof sizeofValue sort split ' +
-        'startsWith stride strideof strideofValue swap toString transcode ' +
-        'underestimateCount unsafeAddressOf unsafeBitCast unsafeDowncast unsafeUnwrap ' +
-        'unsafeReflect withExtendedLifetime withObjectAtPlusZero withUnsafePointer ' +
-        'withUnsafePointerToObject withUnsafeMutablePointer withUnsafeMutablePointers ' +
-        'withUnsafePointer withUnsafePointers withVaList zip'
-    };
+  // https://docs.swift.org/swift-book/ReferenceManual/LexicalStructure.html#ID413
+  // https://docs.swift.org/swift-book/ReferenceManual/zzSummaryOfTheGrammar.html
+  const DOT_KEYWORD = {
+    begin: concat(/\./, either(...Swift.dotKeywords, ...Swift.optionalDotKeywords)),
+    returnBegin: true,
+    contains: [
+      {
+        begin: /\./
+      },
+      {
+        className: 'keyword',
+        begin: either(...Swift.dotKeywords, ...Swift.optionalDotKeywords)
+      }
+    ]
+  };
```

Comment:
```suggestion
  const DOT_KEYWORD = {
    begin: concat(/\./, either(...Swift.dotKeywords, ...Swift.optionalDotKeywords)),
    returnBegin: true,
    relevance: 0,
    contains: [
      {
        begin: /\./,
        relevance: 0
      },
      {
        className: 'keyword',
        begin: either(...Swift.dotKeywords, ...Swift.optionalDotKeywords)
      }
    ]
  };
```

The is a double/triple relevance issue... when building complex rules that really only match a "single" item you need to tune their reelvance such that it is only 1, not 2 or 3.

Reply:
Why do we need to match the `.` here at all though?  Are you just doing that to avoid the parent mode ending?

Reply:
Actually here I think I'd use the top rule to match `.lookahead(blah)` and then just use the inner contains to match the literal `blah`... that should simplify the whole thing.

Reply:
The issue here was that I got a "0 width" error if I didn't provide a rule to match the dot. Is that because I'm using `returnBegin`? My assumption was that whatever I returned also had to be matched by submodes.

W.r.t the relevance, just so I understand it correctly: Is the issue here that both the outer rule **and** the inner rule add to the relevance, and because they're nested, that relevance can be boosted too much?

Reply:
> W.r.t the relevance, just so I understand it correctly: Is the issue here that both the outer rule and the inner rule add to the relevance, and because they're nested, that relevance can be boosted too much?

Yes.  Sometimes one might want this but for matching "simple" things where you're only using rules to jump thru hoops, you do not.  Typically identifiable "things" in a grammar should always score 1, not more than 1.

---

See my latest thoughts above on how to do this without all these hoops.